skbio/metadata/_interval.py

```diff
@@ -0,0 +1,334 @@
+            A list of Interval objects satisfying the search criteria.
+
+        Examples
+        --------
+        >>> from skbio.metadata import IntervalMetadata
+        >>> interval_metadata = IntervalMetadata()
+        >>> interval_metadata.add(intervals=[(0, 2), (4, 7)],
+        ...                       boundaries=None, metadata={'name': 'sagA'})
+        >>> interval_metadata.add(intervals=[(40, 70)],
+        ...                       boundaries=None, metadata={'name': 'sagB'})
+        >>> interval_metadata.query(intervals=[(1, 2)])
+        [Interval(intervals=[(0, 2), (4, 7)], metadata={'name': 'sagA'})]
+
+        Note
+        ----
+        There are two types of queries to perform
+        1. Query by interval.
+        2. Query by key/val pair (i.e. gene=sagA).
+
+        """
+        if self._is_stale_tree:
+            self._rebuild_tree(self._metadata)
+            self._is_stale_tree = False
+
+        invs = set()
+
+        # Find queries by interval
+        if intervals is not None:
+            for value in intervals:
+                invs.update(self._query_interval(value))
+
+        # Find queries by feature attribute
+        if len(invs) == 0 and metadata is not None:
+            invs = set(self._metadata)
+
+        if metadata is not None:
+            invs = self._query_attribute(invs, metadata)
+        return list(invs)
+
+    def drop(self, intervals=None, boundaries=None, metadata=None):
+        """ Drops Interval objects according to a specified query.
+
+        Drops all Interval objects that matches the query.
+
+        Parameters
+        ----------
+        intervals : iterable of tuple of ints
+            A list of intervals associated with the Interval object.
+        boundaries : iterable of tuple of bool
+            A list of boundaries associated with the Interval object.
+        metadata : dict
+            A dictionary of key word attributes associated with the
+            Interval object.
+
+        Examples
+        --------
+        >>> from skbio.metadata import IntervalMetadata
+        >>> interval_metadata = IntervalMetadata()
+        >>> interval_metadata.add(intervals=[(0, 2), (4, 7)],
+        ...                       boundaries=None, metadata={'name': 'sagA'})
+        >>> interval_metadata.add(intervals=[(40, 70)],
+        ...                       boundaries=None, metadata={'name': 'sagB'})
+        >>> interval_metadata.drop(metadata={'name': 'sagA'})
+        >>> interval_metadata.query(metadata={'name': 'sagA'})
+        []
+        """
+        if intervals is None:
+            intervals = []
+        if metadata is None:
+            metadata = {}
+
+        queried_invs = self.query(intervals=intervals,
+                                  boundaries=boundaries,
+                                  metadata=metadata)
+        new_invs = []
+        # iterate through queries and drop them
+        for inv in self._metadata:
+            if inv not in queried_invs:
```

Comment:
I believe this is a list look up. `set` operations here would be really nice although how to do that is not clear right now because of the hashable discussion above